### PR TITLE
Fix queue size calculation to include whole size of tail pages

### PR DESCRIPTION
**Pull request by kaisecheng**
## Release notes

Fix persistent queue size calculation to include whole size of tail pages

## What does this PR do?

This commit uses the capacity of tail pages to do size calculation. 
Every page is initialized with its full capacity, default 64mb. Prior to the change, it counts the size of the events per tail page and the head page. Some pages may have 100kb event, some may have 30mb. Tail pages are immutable and consume a fixed size on disk. So, the size calculation should take the capacity not the event size. Logstash use this size to determine if the queue is full.

## Why is it important/What is the impact to the user?

User relies on queue.max_bytes to control the total queue size. Under the old calculation, queue size can exceed the setting. 
This also changes the metrics in monitoring API /_node/stats `pipelines.main.queue.queue_size_in_bytes`

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## Related issues

- Closes: #14566

### CODE REVIEW

File: logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java [hunk 1]
```diff
@@ -806,7 +806,7 @@ public void getsPersistedByteSizeCorrectly() throws Exception {
                 seqNum = queue.write(new StringElement("foooo"));
             }
             queue.ensurePersistedUpto(seqNum);
-            assertThat(queue.getPersistedByteSize(), is(1063L));
+            assertThat(queue.getPersistedByteSize(), is(1243L));
```
[the remark this answers is not in the document]
Comment (andsel) [reply logstash-core/src/test/java/org/logstash/ackedqueue/QueueTest.java:809]: I think that the `1243` sounds like a magic number.
Please add a comment that describe the above explanation or extract in a variable like:
```ruby
expected_size = num_pages * page_size + head_page_size
```

**Review by andsel — CHANGES_REQUESTED** [review]
The changes is good however left a note a provide better meaning for something that seems like a magic number

**Comment by kaisecheng** [comment]
Abandoned this PR as it has different assumption on how the queue size should calculate.

In the doc, [garbage collection](https://www.elastic.co/guide/en/logstash/current/persistent-queues.html#garbage-collection), `Each page containing unprocessed events will count against the queue.max_bytes byte size.`, while this PR counts total pages size with their max capacity. 

The issue of PQ could exceed queue.max_bytes is still here. We can keep the current calculation for queue.max_bytes (only count unprocessed events) and add a function to ensure the total bytes on disk respects the limit max bytes

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 2]
```diff
@@ -140,7 +140,7 @@ public long getPersistedByteSize() {
                 size = 0L;
             } else {
                 size = headPage.getPageIO().getHead()
-                    + tailPages.stream().mapToLong(p -> p.getPageIO().getHead()).sum();
+                    + tailPages.stream().mapToLong(p -> p.getPageIO().getCapacity()).sum();
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java:143]: The PR was on hold as I noticed the change will break assumption of the metrics related to queue size, and was not clear if we have well defined how max queue size should calculate.

Close this and will find another way to fix the issue.

**Pull request closed without merging**